pkg/server/server.go

```diff
@@ -363,11 +414,25 @@ func (s *Server) Check(ctx context.Context, req *openfgav1.CheckRequest) (*openf
 
 	ctx = typesystem.ContextWithTypesystem(ctx, typesys)
 
-	checkResolver := graph.NewLocalChecker(
-		storagewrappers.NewCombinedTupleReader(s.datastore, req.ContextualTuples.GetTupleKeys()),
-		graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
-		graph.WithMaxConcurrentReads(s.maxConcurrentReadsForCheck),
-	)
+	var checkResolver graph.CheckResolver
+	if s.checkCache == nil {
+		checkResolver = graph.NewLocalChecker(
+			storagewrappers.NewCombinedTupleReader(s.datastore, req.ContextualTuples.GetTupleKeys()),
+			graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
+			graph.WithMaxConcurrentReads(s.maxConcurrentReadsForCheck),
+		)
+	} else {
+		checkResolver = graph.NewLocalChecker(
+			storagewrappers.NewCombinedTupleReader(s.datastore, req.ContextualTuples.GetTupleKeys()),
+			graph.WithResolveNodeBreadthLimit(s.resolveNodeBreadthLimit),
+			graph.WithMaxConcurrentReads(s.maxConcurrentReadsForCheck),
+			graph.WithCachedResolver(
+				graph.WithExistingCache(s.checkCache),
+				graph.WithCacheTTL(s.checkQueryCacheTTL),
+			),
+		)
+	}
+	defer checkResolver.Close()
```

Comment:
This is too verbose and it's duplicate in two places.

I have a suggestion that builds upon your PR: https://gist.github.com/miparnisari/d833b97ba94e8d57ae50e6e8404ca4d6

The gist: we create all the necessary Check options on server startup, and then we use those options in Check and ListObjects

Reply:
I have taken the idea and update the. code. I could not take the code as is as it has a bug in setting incorrect resolve node breadth limit for list object (using maxConcurrentReadsForCheck instead of maxConcurrentReadsForListObjects).